File: 프로그래머스/2/1844. 게임 맵 최단거리/게임 맵 최단거리.java
import java.util.*;

class Solution {
    public int solution(int[][] maps) {
        int answer = 0;
        Queue<int[]> queue = new LinkedList<>();
        
        int n = maps.length;
        int m = maps[0].length;
        int[] dx = {-1, 1, 0, 0};
        int[] dy = {0, 0, -1, 1};
        
        boolean[][] visited = new boolean[n][m];
        queue.offer(new int[]{0, 0, 1});
        visited[0][0] = true;
        
        while (!queue.isEmpty()) {
            int[] current = queue.poll();
            int x = current[0];
            int y = current[1];
            int distance = current[2];
            
            if (x == n-1 && y == m-1) {
                return distance;
            }
            
            for (int i = 0; i < 4; i++) {
                int nx = x + dx[i];
                int ny = y + dy[i];
                if (
                    nx >= 0 && nx < n && ny >= 0 && ny < m && maps[nx][ny] == 1 && !visited[nx][ny]
                ) {
                    queue.offer(new int[]{nx, ny, distance + 1});
                    visited[nx][ny] = true;
                }
            }
        }
        return -1;
    }
}

File: 프로그래머스/2/1844. 게임 맵 최단거리/게임 맵 최단거리.py
def solution(maps):
    answer = 0
    dx = [1, 0, 0, -1]
    dy = [0, 1, -1, 0]
    n = len(maps)
    m = len(maps[0])
    visited = [[False] * m for _ in range(n)]
    queue = [(0, 0, 1)]
    
    while queue:
        x, y, dist = queue.pop(0)
        if x == n-1 and y == m - 1:
                return dist
            
        for i in range(4):
            mx = x + dx[i]
            my = y + dy[i]
            
            if 0 <= mx < n and 0 <= my < m and not visited[mx][my]:
                if maps[mx][my] == 1:
                    visited[mx][my] = True
                    queue.append((mx, my, dist + 1))
    
    return -1